hail/python/hailtop/batch/contrib/regenie/regenie-batch.py

```diff
@@ -0,0 +1,293 @@
+    parser.add_argument('--lowmem', required=False, action='store_true')
+
+    parser.add_argument('--lowmem-prefix', required=False)
+    parser.add_argument('--threads', required=False, default=1)
+
+    if not hexists(path_or_str):
+        print(f"Couldn't find a file named {path_or_str}, assuming this is an argument string")
+        r = parser.parse_args(shlex.split(path_or_str))
+
+    else:
+        print(f"Found {path_or_str}, reading")
+
+        with hopen(path_or_str, "r") as f:
+            t = shlex.split(f.read())
+            r = parser.parse_args(t)
+
+    ninputs = sum(i is not None for i in [r.bed, r.bgen, r.pgen])
+    if ninputs == 0:
+        raise Exception("An input file is required. Please specify: --bed or --bgen or --pgen")
+
+    if ninputs > 1:
+        raise Exception(f"""More than one input file specified.
+                            Check that only one of --bed or --bgen or --pgen are specified for --step {r.step}
+                        """)
+
+    if r.step == 2:
+        if r.pred:
+            print("--pred provided for --step 2, but Batch will constrain the --pred output prefix to that of --step 1.")
+
+        if r.extract or r.exclude:
+            raise Exception("--extract and --exclude only work with --step 1")
+
+    if r.lowmem and not r.lowmem_prefix:
+        raise Exception("When --lowmem provided, --lowmem-prefix required")
+
+    return r
+
+
+# TODO: check --phenoCol, --phenoColList, --remove
+def get_phenos(step_args: argparse.Namespace):
+    with hopen(step_args.phenoFile, "r") as f:
+        return f.readline().strip().split(" ")[2:]
+
+
+def get_input(batch, step_args: argparse.Namespace):
+    add = {}
+    for name, val in vars(step_args).items():
+        if name in from_underscore:
+            name = from_underscore[name]
+
+        if name not in input_file_args or val is None:
+            continue
+
+        if name == "bed":
+            prefix = step_args.bed
+            inf = {}
+            inf["bed"] = f"{prefix}.bed"
+            inf["bim"] = f'{prefix}.bim'
+            inf["fam"] = f'{prefix}.fam'
+            add[name] = batch.read_input_group(**inf)
+            continue
+
+        if name == "pgen":
+            prefix = step_args.pgen
+            inf = {}
+            inf["pgen"] = f'{prefix}.pgen'
+            inf["pvar"] = f'{prefix}.pvar'
+            inf["psam"] = f'{prefix}.psam'
+            add[name] = batch.read_input_group(**inf)
+            continue
+
+        add[name] = batch.read_input(val)
+
+    return add
+
+
+def prepare_jobs(batch, args: BatchArgs, step1_args: argparse.Namespace,
+                 step2_args: argparse.Namespace):
+    j1 = batch.new_job(name='run-regenie')
+    j1.image('akotlar/regenie:latest')
+    j1.cpu(args.cores)
+    j1.memory(args.memory)
+    j1.storage(args.storage)
+
+    in_step1 = get_input(batch, step1_args)
+
+    phenos = get_phenos(step1_args)
+    nphenos = len(phenos)
+
+    s1pre = step1_args.out
```

Comment:
can we call this `step1_output_prefix`?

Reply:
Allowing the user to specify the output file prefix used internally in batch seems unnecessary and appears to add complexity to this implementation?

I guess if we already have it, we can leave it. In the controlled docker environment, I'm inclined to just pick a descriptive name, like `output`. That'll make all this batch resource management code a bit easier to read and removes some argument parsing logic. This all simplifies to:

```python
j1.declare_resource_group(
    output={
        'log': '{root}.log',
        'pred_list': '{root_pred.list}',
        **{f'pheno_{i+1}': '{{root}}_{i+1}.loco' for i in range(0, nphenos)},
        # ... and add lowmem and w/e is necessary
})
```

Reply:
n.b. the `{{root}}` thing becomes `{root}` after python string formatting, then batch further refines that to the right root for this resource group.

Reply:
So I found this frustrating, but maybe I did the wrong thing. The arguments/interface for regenie takes an --out for step1 and --step2. If I want regenie to be able to use these (and why not, they're provided, I shouldn't have to do anything to fight batch on this). The natural thing is for the user to continue using the regenie arguments, and for us to use them as well for the final output.

Reply:
> Allowing the user to specify the output file prefix used internally in batch seems unnecessary and appears to add complexity to this implementation?

Regenie expects --out to be respected, so we need to specify this or clobber --out